src/main/scala/chisel3/util/experimental/LoadMemoryTransform.scala

```diff
@@ -0,0 +1,231 @@
+  * Example (a file containing the decimal values 0, 7, 14, 21):
+  * {{{
+  *   0
+  *   7
+  *   d
+  *  15
+  * }}}
+  * Binary file is similarly constructed.
+  *
+  * ==More info==
+  * See the LoadMemoryFromFileSpec.scala in the test suite for more examples
+  * @see <a href="https://github.com/freechipsproject/chisel3/wiki/Chisel-Memories">Load Memories in the Chisel wiki</a>
+  *
+  * @param target        memory to load
+  * @param fileName      name of input file
+  * @param hexOrBinary   use $readmemh or $readmemb, i.e. hex or binary text input, default is hex
+  */
+case class ChiselLoadMemoryAnnotation(
+  target:      InstanceId,
+  fileName:    String,
+  hexOrBinary: MemoryLoadFileType.FileType = MemoryLoadFileType.Hex
+)
+  extends ChiselAnnotation with RunFirrtlTransform {
+
+  if(fileName.isEmpty) {
+    throw new Exception(
+      s"""LoadMemory from file annotations file empty file name"""
+    )
+  }
+
+  def transformClass: Class[LoadMemoryTransform] = classOf[LoadMemoryTransform]
+
+  def toFirrtl: LoadMemoryAnnotation = {
+    LoadMemoryAnnotation(target.toNamed.asInstanceOf[ComponentName], fileName, hexOrBinary)
+  }
+}
+
+
+object loadMemoryFromFile {
+  def apply(
+    memory: MemBase[_],
+    fileName: String,
+    hexOrBinary: MemoryLoadFileType.FileType = MemoryLoadFileType.Hex
+  ): Unit = {
+    annotate(ChiselLoadMemoryAnnotation(memory, fileName))
+  }
+}
+
+/**
+  * This transform only is activated if verilog is being generated
+  * (determined by presence of the proper emit annotation)
+  * when activated it creates additional verilog files that contain
+  * modules bound to the modules that contain an initializable memory
+  *
+  * Currently the only non-verilog based simulation that can support loading
+  * memory from a file is treadle but it does not need this transform
+  * to do that.
+  */
+//noinspection ScalaStyle
+class LoadMemoryTransform extends Transform {
+  def inputForm: CircuitForm  = LowForm
+  def outputForm: CircuitForm = LowForm
+
+  private var memoryCounter: Int = -1
+
+  private val bindModules: mutable.ArrayBuffer[BlackBoxInlineAnno] = new mutable.ArrayBuffer()
+
+  private val verilogEmitter:    VerilogEmitter = new VerilogEmitter
+
+  /**
+    * run the pass
+    * @param circuit the circuit
+    * @param annotations all the annotations
+    * @return
+    */
+  def run(circuit: Circuit, annotations: AnnotationSeq): Circuit = {
+    val groups = annotations
+      .collect{ case m: LoadMemoryAnnotation => m }
+      .groupBy(_.target.serialize)
+    val memoryAnnotations = groups.map { case (key, annos) =>
+        if (annos.size > 1) {
+          throw new Exception(s"Multiple (${annos.size} found for memory $key one LoadMemoryAnnotation is allowed per memory")
+        }
+        key -> annos.head
+      }
+
+    val modulesByName = {
+      circuit.modules.collect { case m: firrtl.ir.Module => m }.map { module => module.name -> module }.toMap
```

Comment:
I still think this is better as just:
circuit.modules.collect { case m: firrtl.ir.Module => m.name -> m }.toMap

Reply:
done